---
title: "04 Neuron DE Analysis"
author: "Jussi Kupari"
date: "Last edited: `r Sys.Date()`"
output: html_notebook
---

```{r, warning=FALSE}
# Load libraries and functions
suppressPackageStartupMessages({
  library(tidyverse)
  library(magrittr)
  library(Seurat)
  library(UpSetR)
  library(patchwork)
  library(enrichR)
  library(readxl)
  library(fcoex)
  library(harmony)
})
R.utils::sourceDirectory(here::here("R", "functions"), modifiedOnly=FALSE)
```

```{r}
# Load neuron lowq data
(neurons <- readRDS(here::here("data", "proc", "neurons.rds")))
```

```{r}
# Keep only c-fiber types for the analysis 
(neurons %<>% 
  set_idents("usoskin_id") %>% 
  subset(idents = c("NP1", "NP2", "NP3", "PEP1", "TRPM8", "Th", "MRGPRB4")))
```

#### DE between control and each timepoint all with celltypes collapsed together
```{r, results='hide'}
# Get DE genes for each comparison
timepoints <-
  neurons@meta.data$timepoint %>%
  str_subset("0$", negate = TRUE) %>%
  unique()

bulk_grid <- expand_grid(timepoints = timepoints, control = "0")

pseudobulk <-
  bulk_grid %>% 
  pmap( 
    ~ neurons %>%
      set_idents("timepoint_days") %>%
      FindMarkers(ident.1 = .x, ident.2 = .y) %>%
      rownames_to_column("gene") %>% 
      filter(p_val_adj < 10^-20)
  ) %>%
  set_names(timepoints)
```

```{r}
# Collect up- an downregulated genes
pseudobulk_down_genes <-
  pseudobulk %>%
  map(~.x %>% filter(avg_log2FC < 0) %>% pull(gene))

pseudobulk_up_genes <-
  pseudobulk %>%
  map(~.x %>% filter(avg_log2FC > 0) %>% pull(gene))
```

```{r}
# Extract gene intersections between timepoints for pseudobulk
pseudobulk_upregulated_intersections <- get_upset_gene_lists(pseudobulk_up_genes)
pseudobulk_downregulated_intersections <- get_upset_gene_lists(pseudobulk_down_genes)
```

#### Differential expression between control and the first 48h for each celltype
```{r, results='hide'}
# Collect celltypes
celltypes <-
  neurons@meta.data$usoskin_id %>%
  unique()

# Set idents to usoskin
neurons %<>% set_idents("usoskin_id")

# Create function
get_treatment_markers <- 
  as_mapper(
    ~neurons %>%
      subset(idents = .x) %>%
      set_idents("timepoint_days") %>% 
      subset(idents = c("0", "0.25", "0.5", "1", "2")) %>%
      set_idents("treatment") %>% 
      FindMarkers(ident.1 = "RA", ident.2 = "Control") %>%
      rownames_to_column("gene") %>% 
      filter(p_val_adj < 10^-5)
  )

# Make it safe
get_safely_treatment_markers <- safely(get_treatment_markers)

# Get markers for each cell type
de_first_48h <-
  celltypes %>% 
  map(get_safely_treatment_markers) %>% 
  set_names(celltypes) %>% 
  map("result") %>% 
  compact()

# Extract upregulated genes
up_de_first_48h_genes <-
  de_first_48h %>%
  map(~.x %>% filter(avg_log2FC > 0) %>% pull(gene))

# Extract downregulated genes
down_de_first_48h_genes <-
  de_first_48h %>%
  map(~.x %>% filter(avg_log2FC < 0) %>% pull(gene))
```

```{r, fig.width=12}
# Plot upset up first 48h
plot_upset(up_de_first_48h_genes, order = c("NP1", "NP2", "NP3", "PEP1", "Th", "TRPM8", "MRGPRB4"))
```

```{r}
first48h_upregulated_intersections <- get_upset_gene_lists(up_de_first_48h_genes)
```

```{r}
# Plot upset down first 48h
plot_upset(down_de_first_48h_genes, order = c("NP1", "NP2", "NP3", "PEP1", "Th", "TRPM8", "MRGPRB4"))
```

```{r}
first48h_downregulated_intersections <- get_upset_gene_lists(down_de_first_48h_genes)
```

#### Differential expression between control and the late state (33d, 63d) for each celltype
```{r, results='hide'}
# Collect celltypes
celltypes <-
  neurons@meta.data$usoskin_id %>%
  unique()

# Create function
get_treatment_markers <- 
  as_mapper(
    ~neurons %>%
      subset(idents = .x) %>%
      set_idents("timepoint_days") %>% 
      subset(idents = c("0", "33", "63")) %>%
      set_idents("treatment") %>% 
      FindMarkers(ident.1 = "RA", ident.2 = "Control") %>%
      rownames_to_column("gene") %>% 
      filter(p_val_adj < 10^-5)
  )

# Make it safe
get_safely_treatment_markers <- safely(get_treatment_markers)

# Get markers for each cell type
de_late <-
  celltypes %>% 
  map(get_safely_treatment_markers) %>% 
  set_names(celltypes) %>% 
  map("result") %>% 
  compact()

# Extract upregulated genes
up_de_late_genes <-
  de_late %>%
  map(~.x %>% filter(avg_log2FC > 0) %>% pull(gene))

# Extract downregulated genes
down_de_late_genes <-
  de_late %>%
  map(~.x %>% filter(avg_log2FC < 0) %>% pull(gene))
```

```{r, fig.width=12}
# Plot upset up late state
plot_upset(up_de_late_genes, order = c("NP1", "NP2", "NP3", "PEP1", "Th", "TRPM8", "MRGPRB4"))
```

```{r}
late_upregulated_intersections <- get_upset_gene_lists(up_de_late_genes)
```

```{r}
# Plot upset down late state
plot_upset(down_de_late_genes, order = c("NP1", "NP2", "NP3", "PEP1", "Th", "TRPM8", "MRGPRB4"))
```

```{r}
late_downregulated_intersections <- get_upset_gene_lists(down_de_late_genes)
```

#### Enrichr analysis for gene sets
```{r}
# Get available DBs to choose from
(dbs_all <- listEnrichrDbs())
```

```{r}
# First 48h regulated NP1/PEP1
np1_pep1_up_early <-
  first48h_upregulated_intersections[["PEP1_NP1"]][["gene"]] %>% 
  get_enriched(dbs_vector = "GO_Biological_Process_2021") 

np1_pep1_up_early <-
np1_pep1_up_early[["GO_Biological_Process_2021"]] %>% 
  filter(Adjusted.P.value < 0.05) 

np1_pep1_down_early <-
  first48h_downregulated_intersections[["PEP1_NP1"]][["gene"]] %>% 
  get_enriched(dbs_vector = "GO_Biological_Process_2021") 

np1_pep1_down_early <-
np1_pep1_down_early[["GO_Biological_Process_2021"]] %>% 
  filter(Adjusted.P.value < 0.05) 
```

```{r}
# late state regulated NP1/PEP1
np1_pep1_up_late <-
  late_upregulated_intersections[["PEP1_NP1"]][["gene"]] %>% 
  get_enriched(dbs_vector = "GO_Biological_Process_2021") 

np1_pep1_up_late <-
np1_pep1_up_late[["GO_Biological_Process_2021"]] %>% 
  filter(Adjusted.P.value < 0.05) 

np1_pep1_down_late <-
  late_downregulated_intersections[["PEP1_NP1"]][["gene"]] %>% 
  get_enriched(dbs_vector = "GO_Biological_Process_2021") 

np1_pep1_down_late <-
np1_pep1_down_late[["GO_Biological_Process_2021"]] %>% 
  filter(Adjusted.P.value < 0.05) 
```

```{r, fig.width=12}
# Plot upregulated NP1/PEP1 intersection for the first 48h
np1_pep1_up_early %>% 
  mutate(
    Term = str_trunc(str_remove(Term, "\\(GO:\\d+\\)"), 70, "right")
    ) %>% 
  filter(Adjusted.P.value < 0.05) %>%
  arrange(desc(Combined.Score)) %>% 
  slice(1:20) %>% 
  ggplot(
    aes(Odds.Ratio, fct_reorder(Term, Odds.Ratio), 
        size = Odds.Ratio, 
        color = Adjusted.P.value)
    ) +
  geom_point() +
  labs(y = "Term") +
  theme_classic()
```

```{r, fig.width=12}
# Plot downregulated NP1/PEP1 intersection for the first 48h
np1_pep1_down_early %>% 
  mutate(
    Term = str_trunc(str_remove(Term, "\\(GO:\\d+\\)"), 70, "right")
    ) %>%
  filter(Adjusted.P.value < 0.05) %>%
  arrange(desc(Combined.Score)) %>% 
  slice(1:20) %>% 
  ggplot(
    aes(Odds.Ratio, fct_reorder(Term, Odds.Ratio), 
        size = Odds.Ratio, 
        color = Adjusted.P.value)
    ) +
  geom_point() +
  labs(y = "Term") +
  theme_classic()
```

```{r, fig.width=12}
# Plot upregulated NP1/PEP1 intersection for the last timepoints
np1_pep1_up_late %>% 
  mutate(
    Term = str_trunc(str_remove(Term, "\\(GO:\\d+\\)"), 70, "right")
    ) %>%
  filter(Adjusted.P.value < 0.05) %>%
  arrange(desc(Combined.Score)) %>% 
  slice(1:20) %>% 
  ggplot(
    aes(Odds.Ratio, fct_reorder(Term, Odds.Ratio), 
        size = Odds.Ratio, 
        color = Adjusted.P.value)
    ) +
  geom_point() +
  labs(y = "Term") +
  theme_classic()
```

```{r, fig.width=12}
# Plot downregulated NP1/PEP1 intersection for the last timepoints
np1_pep1_down_late %>% 
  mutate(
    Term = str_trunc(str_remove(Term, "\\(GO:\\d+\\)"), 70, "right")
    ) %>%
  filter(Adjusted.P.value < 0.05) %>%
  arrange(desc(Combined.Score)) %>% 
  slice(1:20) %>% 
  ggplot(
    aes(Odds.Ratio, fct_reorder(Term, Odds.Ratio), 
        size = Odds.Ratio, 
        color = Adjusted.P.value)
    ) +
  geom_point() +
  labs(y = "Term") +
  theme_classic()
```

#### Downsample data for fcoex
```{r}
# Downsample data to smaller size (to avoid memory issues with fcoex)
set.seed(1)

downsampled_cells <-
  neurons@meta.data %>%
  rownames_to_column("cell") %>% 
  group_by(usoskin_id) %>% 
  slice_sample(prop = 0.25) %>% 
  pull("cell")

(small_data <-
  neurons %>% 
  subset(cells = downsampled_cells))
```

```{r, fig.width=12}
# Check downsampled data
(DimPlot(neurons, 
         shuffle = TRUE, 
         label = TRUE, 
         repel = TRUE,
         group.by = "usoskin_id") + 
   NoLegend()) +
  
(DimPlot(small_data, 
         shuffle = TRUE, 
         label = TRUE,
         repel = TRUE,
         group.by = "usoskin_id") + 
   NoLegend()) +
  
   (DimPlot(small_data, 
            shuffle = TRUE, 
            group.by = "treatment")) 
```

#### FCOEX using all celltypes together and timepoint as target
```{r, results='hide'}
# Run Fcoex
small_data %<>% set_idents("timepoint_days") 

fc_full <-
  small_data %>% 
  run_fcoex2(n_genes = 250)
```

```{r}
# Wrangle fcoex results
mod_genes <- fc_full@module_list

up_genes_total <-
  pseudobulk_up_genes %>% 
  reduce(union)

down_genes_total <-
  pseudobulk_down_genes %>% 
  reduce(union)

up_and_down <- intersect(up_genes_total, down_genes_total)

genes_to_keep <- setdiff(up_genes_total, up_and_down)

mods <-
mod_genes %>% 
  map(~ .x %>% keep(.x %in% genes_to_keep)) %>% 
  keep(~ length(.x) > 5) %>% 
  map(sort)
```

```{r}
# Plot upset
plot_upset(mods, order = names(mods))
```

```{r, fig.width=12, results='hide'}
# Heatmap of average Ifit1 module gene expression across timepoints
neurons %>% 
  set_idents("timepoint_days") %>%
  ScaleData(features = rownames(neurons)) %>%
  AverageExpression(return.seurat = TRUE) %>% 
  set_cluster_order(c(0, 0.25, 0.5, 1, 2, 12, 33, 63)) %>% 
  DoHeatmap(features = mods$Ifit1, draw.lines = FALSE) 
```

```{r, fig.height=12, results='hide'}
# Plot Ifit1 module for all celltypes individually
plot_modules_ifit1("NP1") +
  plot_modules_ifit1("NP2") +
  plot_modules_ifit1("NP3") +
  plot_modules_ifit1("Th") +
  plot_modules_ifit1("PEP1") +
  plot_modules_ifit1("TRPM8") +
  plot_modules_ifit1("MRGPRB4")
```

```{r, fig.height=12, results='hide'}
# Heatmap of average Apoe module gene expression across timepoints
neurons %>% 
  set_idents("timepoint_days") %>%
  ScaleData(features = rownames(neurons)) %>%
  AverageExpression(return.seurat = TRUE) %>% 
  set_cluster_order(c(0, 0.25, 0.5, 1, 2, 12, 33, 63)) %>% 
  DoHeatmap(features = mods$Apoe, draw.lines = FALSE) 
```

```{r, fig.height=12, results='hide'}
# Plot Apoe module for all celltypes individually
plot_modules_apoe("NP1") +
  plot_modules_apoe("NP2") +
  plot_modules_apoe("NP3") +
  plot_modules_apoe("Th") +
  plot_modules_apoe("PEP1") +
  plot_modules_apoe("TRPM8") +
  plot_modules_apoe("MRGPRB4")
```

```{r}
# Save module object
saveRDS(fc_full, file = here::here("data", "proc", "fcoex_full_250genes.rds"))
```

```{r}
# Run enrichr
go_modules <-
  mod_genes %>% 
  map(~ .x %>% keep(.x %in% genes_to_keep)) %>% 
  keep(~ length(.x) > 5) %>% 
  map(get_enriched, dbs_vector = "GO_Biological_Process_2021")
```

```{r}
# Remove redundant terms using Revigo
# http://revigo.irb.hr/

# go_modules[[1]][[1]] %>% 
#   mutate(GO = str_extract(Term, "GO:\\d+")) %>%
#   mutate(combo = paste(GO, Adjusted.P.value, sep = " ")) %>% 
#   pull(combo) %>% print_gene_list_from_vector() <- copy/paste to the website

revigo_ifit1 <- 
  read.delim("/crex/proj/uppstore2017283/jussi/Su2022/revigo_ifit1_small_new.tsv") %>% 
  filter(Eliminated == "False")

revigo_apoe <- 
  read.delim("/crex/proj/uppstore2017283/jussi/Su2022/revigo_apoe_small_new.tsv") %>% 
  filter(Eliminated == "False")

go_modules[[1]][[1]] %<>% 
  mutate(GO = str_extract(Term, "GO:\\d+")) %>%
  mutate(combo = paste(GO, Adjusted.P.value, sep = " ")) %>% 
  semi_join(revigo_ifit1, by = c("GO" = "TermID"))

go_modules[[2]][[1]] %<>% 
  mutate(GO = str_extract(Term, "GO:\\d+")) %>%
  mutate(combo = paste(GO, Adjusted.P.value, sep = " ")) %>% 
  semi_join(revigo_apoe, by = c("GO" = "TermID"))
```

```{r, fig.height=6, fig.width=12}
# Plot GOs for Ifit1 module
go_modules[[1]][[1]] %>% 
  mutate(
    Term = str_trunc(str_remove(Term, "\\(GO:\\d+\\)"), 70, "right")
    ) %>%
  filter(Adjusted.P.value < 0.05) %>%
  arrange(desc(Combined.Score)) %>% 
  slice(1:20) %>% 
  ggplot(
    aes(Odds.Ratio, fct_reorder(Term, Odds.Ratio), 
        size = Odds.Ratio, 
        color = Adjusted.P.value)
    ) +
  geom_point() +
  labs(y = "Term") +
  theme_classic()
```

```{r, fig.height=6, fig.width=12}
# Plot GOs for Apoe module
go_modules[[2]][[1]] %>% 
  mutate(
    Term = str_trunc(str_remove(Term, "\\(GO:\\d+\\)"), 70, "right")
    ) %>%
  filter(Adjusted.P.value < 0.05) %>%
  arrange(desc(Combined.Score)) %>% 
  slice(1:20) %>% 
  ggplot(
    aes(Odds.Ratio, fct_reorder(Term, Odds.Ratio), 
        size = Odds.Ratio, 
        color = Adjusted.P.value)
    ) +
  geom_point() +
  labs(y = "Term") +
  theme_classic()
```

```{r}
# Create module scores
for (i in seq_along(mods)) {
  neurons %<>% 
    AddModuleScore(features = list(mods[[i]]), name = paste0(names(mods[i]), "_"))
}
```

```{r, fig.width=12}
# Plot module scores for Ifit1
neurons@meta.data %>% 
  ggplot(aes(factor(timepoint_days), Ifit1_1, fill = usoskin_id)) +
  geom_boxplot(outlier.shape = NA) +
  facet_wrap(vars(usoskin_id)) +
  cowplot::theme_cowplot() +
  labs(x = "Timepoint in days", y = "Gene module score", fill = "Neuron type")
```

```{r, fig.width=12}
# Plot module scores for Apoe
neurons@meta.data %>% 
  ggplot(aes(factor(timepoint_days), Apoe_1, fill = usoskin_id)) +
  geom_boxplot(outlier.shape = NA) +
  facet_wrap(vars(usoskin_id)) +
  cowplot::theme_cowplot() +
  labs(x = "Timepoint", y = "Module score", fill = "Neuron type")
```

```{r, fig.width=12}
# Plot two highly induced genes that are not part of the Ifit1 module
VlnPlot(neurons, c("Lcn2", "Saa3"), group.by = "timepoint_days", pt.size = 0, log = TRUE)
```

#### Augur: Perturbation analysis
```{r}
# Add needed columns to metadata for running Augur
neurons@meta.data %<>%
  mutate(
    cell_type = as.character(neurons@meta.data$usoskin_id),
    label = factor(timepoint_days))
```

```{r}
# Get a list of mouse protein coding genes excluding Y chromosome

# https://www.ensembl.org/biomart/martview/766d94f98a348ad7b78782227f808ad4
mouse_genes <- readRDS(here::here("data", "ref", "mouse_genes_ensembl.rds"))

genes_use <-
  mouse_genes  %>% 
  filter(
    Gene.name != "",
    Gene.type == "protein_coding",
    Chromosome.scaffold.name != "Y"
  )  %>% 
  pull(Gene.name)
```

```{r}
# Filter data and create a new Seurat object
neurons2 <- as.matrix(GetAssayData(neurons, slot = "counts"))
neurons2 <- neurons2[row.names(neurons) %in% genes_use, ]

neurons2 %<>%  
  CreateSeuratObject() %>% 
  NormalizeData() %>% 
  FindVariableFeatures() %>% 
  ScaleData() %>% 
  RunPCA(npcs = 100, verbose = TRUE)

neurons2@meta.data <-
  neurons2@meta.data %>% 
  select(-orig.ident) %>% 
  bind_cols(
    neurons@meta.data %>% 
      select(label, cell_type, orig.ident, treatment)
  )
```

```{r}
# Check, ok
neurons2@meta.data %>% select(label, cell_type, treatment)
```

```{r}
# Split object and create combinations 
split_neurons <- SplitObject(neurons2, "label")

grid <- expand_grid(ctrl = "0", ra = c("0.25", "0.5", "1", "2", "12", "33", "63"))

combos <-
  grid  %>% 
  pmap(~ merge(split_neurons[[.x]], split_neurons[[.y]])) %>% 
  set_names(grid$ra)
```

```{r}
# Test augur with one timepoint
augur <- Augur::calculate_auc(combos[["0.25"]], n_threads = parallel::detectCores() - 1)
```

```{r}
# Check
augur$AUC
```

```{r, results='hide'}
# Run Augur for all combinations
augur_combos <-
  combos %>%
  map(Augur::calculate_auc, n_threads = parallel::detectCores() - 1)
```

```{r}
# Save Augur results
saveRDS(augur_combos, file = here::here("data", "proc", "augur_combos.rds"))
```

```{r}
# Wrangle augur results for plotting
aucs <-
  augur_combos %>%
  map("AUC")

aucs %<>%
  map2(names(aucs), ~ {
    .x %<>% mutate(time = .y)
    return(.x)})

aucs %<>%
  bind_rows() %>%
  mutate(time = fct_inorder(time))

aucs %<>%
  split(.$cell_type)
```

```{r, fig.width=12}
# Plot Augur scores 
p <-
  plot_aucs(aucs$NP1) +
  plot_aucs(aucs$NP2) +
  plot_aucs(aucs$MRGPRB4) +
  plot_aucs(aucs$NP3) +
  plot_aucs(aucs$PEP1) +
  plot_aucs(aucs$Th) +
  plot_aucs(aucs$TRPM8) 
 
gt <- patchwork::patchworkGrob(p)
gridExtra::grid.arrange(gt, left = "Timepoint days", bottom = "Perturbation Score")
```

```{r}
# Sessioninfo
sessionInfo()
```

```{r}
# Done
clear_libraries()
```
